src/ciphers/idea.c

```diff
@@ -0,0 +1,252 @@
+ */
+
+#include "tomcrypt.h"
+
+#ifdef LTC_IDEA
+
+const struct ltc_cipher_descriptor idea_desc = {
+   "idea",
+   24,                  /* cipher_ID */
+   16, 16, 8, 8,        /* min_key_len, max_key_len, block_len, default_rounds */
+   &idea_setup,
+   &idea_ecb_encrypt,
+   &idea_ecb_decrypt,
+   &idea_test,
+   &idea_done,
+   &idea_keysize,
+   NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL, NULL
+};
+
+typedef unsigned short int ushort16;
+
+#define _LOW16(x)     ((x)&0xffff)  /* compiler should be able to optimize this away if x is 16 bits */
+#define _HIGH16(x)    ((x)>>16)
+#define _MUL(a,b)     {                                               \
+                         ulong32 p = (ulong32)_LOW16(a) * b;          \
+                         if (p) {                                     \
+                            p = _LOW16(p) - _HIGH16(p);               \
+                            a = (ushort16)p - (ushort16)_HIGH16(p);   \
+                         }                                            \
+                         else                                         \
+                            a = 1 - a - b;                            \
+                      }
+#define _STORE16(x,y) { (y)[0] = (unsigned char)(((x)>>8)&255); (y)[1] = (unsigned char)((x)&255); }
+#define _LOAD16(x,y)  { x = ((ushort16)((y)[0] & 255)<<8) | ((ushort16)((y)[1] & 255)); }
+
+static ushort16 _mul_inv(ushort16 x)
+{
+   ushort16 y = x;
+   unsigned i;
+
+   for (i = 0; i < 15; i++) {
+      _MUL(y, _LOW16(y));
+      _MUL(y, x);
+   }
+   return _LOW16(y);
+}
+
+static ushort16 _add_inv(ushort16 x)
+{
+   return _LOW16(0 - x);
+}
+
+static int _setup_key(const unsigned char *key, symmetric_key *skey)
+{
+   int i, j;
+   ushort16 *e_key = skey->idea.ek;
+   ushort16 *d_key = skey->idea.dk;
+
+   /* prepare enc key */
+   for (i = 0; i < 8; i++) {
+      _LOAD16(e_key[i], key + 2 * i);
+   }
+   for (; i < LTC_IDEA_KEYLEN; i++) {
+      j = (i - i % 8) - 8;
+      e_key[i] = _LOW16((e_key[j+(i+1)%8] << 9) | (e_key[j+(i+2)%8] >> 7));
+   }
+
+   /* prepare dec key */
+   for (i = 0; i < LTC_IDEA_ROUNDS; i++) {
+      d_key[i*6+0] = _mul_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+0]);
+      d_key[i*6+1] = _add_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+1+(i>0 ? 1 : 0)]);
+      d_key[i*6+2] = _add_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+2-(i>0 ? 1 : 0)]);
+      d_key[i*6+3] = _mul_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+3]);
+      d_key[i*6+4] =          e_key[(LTC_IDEA_ROUNDS-1-i)*6+4];
+      d_key[i*6+5] =          e_key[(LTC_IDEA_ROUNDS-1-i)*6+5];
+   }
+   d_key[i*6+0] = _mul_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+0]);
+   d_key[i*6+1] = _add_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+1]);
+   d_key[i*6+2] = _add_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+2]);
+   d_key[i*6+3] = _mul_inv(e_key[(LTC_IDEA_ROUNDS-i)*6+3]);
+
+   return CRYPT_OK;
+}
+
+static int _process_block(const unsigned char *in, unsigned char *out, ushort16 *m_key)
```

Comment:
should we implement `LTC_CLEAN_STACK` variants immediately?

Reply:
see ee484a0bc0